### PR TITLE
feat(behavior_path): use multi thread

### PR DESCRIPTION
**Pull request by tkimura4**
Signed-off-by: tomoya.kimura <tomoya.kimura@tier4.jp>

## Description
To solve https://github.com/autowarefoundation/autoware.universe/issues/275 in behavior path planner, multi thread is used for callbacks by passing a different callback group, whose type is MutuallyExclusive, to each calback.

## Related links
https://github.com/autowarefoundation/autoware.universe/issues/275
https://github.com/autowarefoundation/autoware.universe/pull/260

## Notes for reviewers
Run simple psim, and see if each callback of subscribes is called in a steady cycle.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/behavior_path_planner_node.cpp [hunk 1]
```diff
@@ -518,11 +553,13 @@ void BehaviorPathPlannerNode::run()
...
 
-  if (planner_data_->parameters.visualize_drivable_area_for_shared_linestrings_lanelet) {
+  if (planner_data->parameters.visualize_drivable_area_for_shared_linestrings_lanelet) {
     const auto drivable_area_lines = marker_utils::createFurthestLineStringMarkerArray(
-      util::getDrivableAreaForAllSharedLinestringLanelets(planner_data_));
+      util::getDrivableAreaForAllSharedLinestringLanelets(planner_data));
     debug_drivable_area_lanelets_publisher_->publish(drivable_area_lines);
   }
+
+  mutex_.unlock();
```
Comment (takayuki5168) [inline planning/behavior_path_planner/src/behavior_path_planner_node.cpp:562]: This has to be removed.

File: planning/behavior_path_planner/src/behavior_path_planner_node.cpp [hunk 2]
```diff
@@ -445,41 +465,56 @@ BehaviorTreeManagerParam BehaviorPathPlannerNode::getBehaviorTreeManagerParam()
...
+  const auto output = bt_manager_->run(planner_data);
 
   // path handling
   const auto path = getPath(output);
   const auto path_candidate = getPathCandidate(output);
+
+  // update planner data
+  mutex_.lock();  // for planner_data_
   planner_data_->prev_output_path = path;
+  mutex_.unlock();
```
Comment (takayuki5168) [inline planning/behavior_path_planner/src/behavior_path_planner_node.cpp:517]: Here, not planner_data_, but planner_data.

Comment (takayuki5168) [reply planning/behavior_path_planner/src/behavior_path_planner_node.cpp:517]: without mutex

File: planning/behavior_path_planner/include/behavior_path_planner/behavior_path_planner_node.hpp [hunk 3]
```diff
@@ -49,6 +49,7 @@
 #include <lanelet2_traffic_rules/TrafficRulesFactory.h>
```
Comment (takayuki5168) [inline planning/behavior_path_planner/include/behavior_path_planner/behavior_path_planner_node.hpp:49]: `planner_data_` is still used in the following functions.
Mutex is required, or we can give a local variable `planner_data` to functions.
getPath
getPathCandidate
publishModuleStatus
updateCurrentPose

File: planning/behavior_path_planner/src/behavior_path_planner_node.cpp [hunk 4]
```diff
@@ -668,10 +709,12 @@ void BehaviorPathPlannerNode::onForceApproval(const PathChangeModule::ConstShare
 }
 void BehaviorPathPlannerNode::onMap(const HADMapBin::ConstSharedPtr msg)
 {
+  std::lock_guard<std::mutex> lock(mutex_);
   planner_data_->route_handler->setMap(*msg);
 }
 void BehaviorPathPlannerNode::onRoute(const HADMapRoute::ConstSharedPtr msg)
 {
+  std::lock_guard<std::mutex> lock(mutex_);
```
Comment (takayuki5168) [inline planning/behavior_path_planner/src/behavior_path_planner_node.cpp:717]: In this function, the modified member variables are not only `planner_data_` but also `bt_manager_`, which are used in other functions as well.

So we have to do mutex for `bt_manager_` too. Also the performance would be better by using the other mutex like `mutex_bt_manager_`. But you can use the same mutex for `planner_data_` and `bt_manager_` if you want to. (I'm not sure how much the performance would be improved)

File: planning/behavior_path_planner/src/behavior_path_planner_node.cpp (hunk 1 above)
Comment (tkimura4) [reply planning/behavior_path_planner/src/behavior_path_planner_node.cpp]: [15caaee](https://github.com/autowarefoundation/autoware.universe/pull/657/commits/15caaeec35c22b5493d7ccbcca6a74211fc17171)

File: planning/behavior_path_planner/include/behavior_path_planner/behavior_path_planner_node.hpp (hunk 3 above)
Comment (tkimura4) [reply planning/behavior_path_planner/include/behavior_path_planner/behavior_path_planner_node.hpp:49]: [b6cd6b8](https://github.com/autowarefoundation/autoware.universe/pull/657/commits/b6cd6b8646ad7708f413d9ce680b39605e2fd689)

File: planning/behavior_path_planner/src/behavior_path_planner_node.cpp (hunk 2 above)
Comment (tkimura4) [reply planning/behavior_path_planner/src/behavior_path_planner_node.cpp]: The lock of mutex is required here for value-assignment to member variables to planner_data_

File: planning/behavior_path_planner/src/behavior_path_planner_node.cpp (hunk 4 above)
Comment (tkimura4) [reply planning/behavior_path_planner/src/behavior_path_planner_node.cpp]: Fixed at [e12262a](https://github.com/autowarefoundation/autoware.universe/pull/657/commits/e12262affa88455946c65b61ad4c0195a1fd8439).

`bt_manager_` is used only by initializer and `run()`

**Review by yukkysaito — DISMISSED** [review]

**Review by TakaHoribe — APPROVED** [review]
It worked fine in psim and the code LGTM too.

**Comment by tkimura4** [comment]
I've resolved the conflict with main, so please re-approve

**Review by takayuki5168 — APPROVED** [review]
LGTM

**Pull request merged**